src/transforms/sample/transform.rs

```diff
@@ -104,19 +133,29 @@ impl FunctionTransform for Sample {
             Event::Metric(_) => panic!("component can never receive metric events"),
         });
 
-        let counter_value: u64 = *self.counter.entry(group_by_key.clone()).or_default();
+        let counter_value: f32 = *self
+            .counter
+            .entry(group_by_key.clone())
+            .or_insert(self.rate.pct());
 
-        let num = if let Some(value) = value {
-            seahash::hash(value.as_bytes())
+        // reset counter for particular key, or default key if group_by option isn't provided
+        let increment: f32 = counter_value + self.rate.pct();
+        self.counter.insert(
+            group_by_key.clone(),
+            if increment >= 1.0 {
+                increment - 1.0
+            } else {
+                increment
+            },
+        );
+
+        let should_process = if let Some(value) = value {
+            (seahash::hash(value.as_bytes()) % 100) as f32 <= (self.rate.pct() * 100.0)
         } else {
-            counter_value
+            increment >= 1.0
```

Comment:
This will never be true because of the calculation on `increment` above. I think you will want to move all of the counter calculation into this `else` block.

Reply:
`increment` is not marked as `mut`, what is inserted into the map above is a new value `increment - 1.0` or `increment`

Reply:
Ah, I totally missed that. Gotcha.